editor/src/messages/portfolio/document/node_graph/node_graph_message_handler.rs

```diff
@@ -397,20 +415,34 @@ impl NodeGraphMessageHandler {
 					warn!("Removing input of invalid node type '{}'", node.name);
 					return false;
 				};
+
 				if let NodeInput::Value { tagged_value, .. } = &node_type.inputs[input_index].default {
-					*input = NodeInput::value(tagged_value.clone(), true);
+					let mut refers_to_output_node = false;
+
+					// use the first input node as the new input if deleting node's first input is a node,
+					// and the current node uses its primary output too
+					if let Some(input_node) = &first_input_node {
+						if *output_index == 0 {
+							refers_to_output_node = true;
+							*input = input_node.clone()
+						}
+					}
+
+					if !refers_to_output_node {
+						*input = NodeInput::value(tagged_value.clone(), true);
+					}
 				}
 			}
 			if let DocumentNodeImplementation::Network(network) = &mut node.implementation {
-				Self::remove_references_from_network(network, deleting_node_id);
+				return Self::remove_references_from_network(network, deleting_node_id, disconnect_nodes);
```

Comment:
I've realised that we don't need any recursion here, and so the if let block can actually be deleted.

Reply:
<img width="700" alt="image" src="https://github.com/GraphiteEditor/Graphite/assets/83733638/74115d4a-842e-42ee-82e4-c1a535e5b4e0">
Removing the if-let block is causing an `out of bounds access` memory error